src/Arbor/File/Format/Asif/Write.hs

```diff
@@ -0,0 +1,183 @@
+{-# LANGUAGE MultiParamTypeClasses #-}
+
+module Arbor.File.Format.Asif.Write
+
+where
+
+import Arbor.File.Format.Asif.ByteString.Builder
+import Arbor.File.Format.Asif.Type
+import Arbor.File.Format.Asif.Whatever           (Whatever (..))
+import Conduit
+import Control.Foldl
+import Control.Lens
+import Control.Monad.IO.Class                    (liftIO)
+import Control.Monad.Trans.Resource              (MonadResource)
+import Data.Int
+import Data.Word
+import System.IO                                 (Handle, SeekMode (AbsoluteSeek), hFlush, hSeek)
+import System.IO.Temp                            (openTempFile)
+
+import qualified Arbor.File.Format.Asif.Format as F
+import qualified Data.ByteString.Builder       as BB
+import qualified Data.ByteString.Lazy          as LBS
+import qualified Data.IP                       as IP
+import qualified Data.Text                     as T
+import qualified Data.Text.Lazy                as TL
+import qualified Data.Text.Lazy.Encoding       as TE
+
+import qualified Data.Thyme.Clock.POSIX as TY
+import qualified Data.Thyme.Time.Core   as TY
+
+writeAsif :: (Foldable f, MonadResource m)
+  => Handle
+  -> String
+  -> Maybe TY.POSIXTime
+  -> FoldM m a [Segment Handle]
+  -> f a
+  -> m ()
+writeAsif hOutput asifType mTimestamp fld foldable = do
+  segments <- foldM fld foldable
+  contents  <- segmentsC asifType mTimestamp segments
+  runConduit $ contents .| sinkHandle hOutput
+  liftIO $ hFlush hOutput
+
+buildAsifBytestring :: (Foldable f, MonadResource m)
+  => String
+  -> Maybe TY.POSIXTime
+  -> FoldM m a [Segment Handle]
+  -> f a
+  -> m LBS.ByteString
+buildAsifBytestring asifType mTimestamp fld foldable = do
+  (_, _, h) <- openTempFile Nothing "asif"
+  writeAsif h asifType mTimestamp fld foldable
+  liftIO $ hSeek h AbsoluteSeek 0
+  liftIO $ LBS.hGetContents h
+
+-----
+
+-- This can in priciple cover all of StringZ, Text, Binary, Bitmap, and Bitstring, as well as unknown encodings.
+-- Correctly encoding the value is the responsibility of the caller.
```

Comment:
I'm not really understanding the comment.  Can you elaborate on what the caller is supposed to do and give an example?